### PR TITLE
Support Harvester CCM and CSI

### PR DESCRIPTION
**Pull request by yaocw2020**
Signed-off-by: yaocw2020 <yaocanwu@gmail.com>

#### Proposed Changes ####

To support harvester cloud provider and harvester csi

#### Linked Issues ####

https://github.com/rancher/rke2/issues/1680
https://github.com/harvester/harvester/issues/1019

### CODE REVIEW

**Comment by gitlawr** [comment]
Should we publish Harvester airgap images tarballs as we did for vSphere provider?

**Comment by dweomer** [comment]
> Should we publish Harvester airgap images tarballs as we did for vSphere provider?

@gitlawr IMO, yes.

**Comment by yaocw2020** [comment]
Dependency: https://github.com/rancher/charts/pull/1426

**Review by dweomer — APPROVED** [review]

File: scripts/build-images [hunk 1]
```diff
@@ -71,5 +71,14 @@ xargs -n1 -t docker image pull --quiet << EOF > build/images-multus.txt
     ${REGISTRY}/rancher/hardened-sriov-network-webhook:v1.0.0-build20210429
 EOF
 
+xargs -n1 -t docker image pull --quiet << EOF > build/images-harvester.txt
+    ${REGISTRY}/rancher/harvester-cloud-provider:v0.1.1
+    ${REGISTRY}/rancher/harvester-csi-driver:v0.1.1
+    ${REGISTRY}/longhornio/csi-node-driver-registrar:v1.2.0-lh1
+    ${REGISTRY}/longhornio/csi-resizer:v0.5.1-lh1
+    ${REGISTRY}/longhornio/csi-provisioner:v1.6.0-lh1
+    ${REGISTRY}/longhornio/csi-attacher:v2.2.1-lh1
```
Comment (brandond) [inline scripts/build-images:80]: All airgapped images need to be in the rancher repo. You might need to get these copied in via a PR to the rancher/image-mirror repo.

**Review by brandond — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 2]
```diff
@@ -105,6 +105,8 @@ RUN CHART_VERSION="2.11.100-build2021022302"  CHART_FILE=/charts/rke2-metrics-se
 RUN CHART_VERSION="v3.7.1-build2021041604"    CHART_FILE=/charts/rke2-multus.yaml         CHART_BOOTSTRAP=true   /charts/build-chart.sh
 RUN CHART_VERSION="1.0.000"                   CHART_FILE=/charts/rancher-vsphere-cpi.yaml CHART_BOOTSTRAP=true   CHART_REPO="https://charts.rancher.io" /charts/build-chart.sh
 RUN CHART_VERSION="2.1.000"                   CHART_FILE=/charts/rancher-vsphere-csi.yaml CHART_BOOTSTRAP=true   CHART_REPO="https://charts.rancher.io" /charts/build-chart.sh
+RUN CHART_VERSION="0.1.1"                     CHART_FILE=/charts/harvester-cloud-provider.yaml CHART_BOOTSTRAP=true   CHART_REPO="https://charts.rancher.io" /charts/build-chart.sh
```
Comment (cjellick) [inline Dockerfile:108]: Do the charts have to be in the rancher repo? Are they used outside of the context of rke2?

We discovered with the vsphere chart that there is a bit of a disconnect between how rancher releases charts and how rke2 consumes them. 

The workflow for updating charts might be better if these charts lived in the rke2-charts repo

Comment (guangbochen) [reply Dockerfile:108]: currently, the PR is submitted against the rancher/charts repo https://github.com/rancher/charts/pull/1426, if rke2-charts is making much more sense we should be able to change that, cc @yaocw2020

File: Dockerfile [hunk 3]
```diff
@@ -105,6 +105,8 @@ RUN CHART_VERSION="2.11.100-build2021022302"  CHART_FILE=/charts/rke2-metrics-se
 RUN CHART_VERSION="v3.7.1-build2021041604"    CHART_FILE=/charts/rke2-multus.yaml         CHART_BOOTSTRAP=true   /charts/build-chart.sh
 RUN CHART_VERSION="1.0.000"                   CHART_FILE=/charts/rancher-vsphere-cpi.yaml CHART_BOOTSTRAP=true   CHART_REPO="https://charts.rancher.io" /charts/build-chart.sh
 RUN CHART_VERSION="2.1.000"                   CHART_FILE=/charts/rancher-vsphere-csi.yaml CHART_BOOTSTRAP=true   CHART_REPO="https://charts.rancher.io" /charts/build-chart.sh
+RUN CHART_VERSION="0.1.100-rc00"              CHART_FILE=/charts/harvester-cloud-provider.yaml CHART_BOOTSTRAP=true /charts/build-chart.sh
+RUN CHART_VERSION="0.1.200-rc00"              CHART_FILE=/charts/harvester-csi-driver.yaml     CHART_BOOTSTRAP=true /charts/build-chart.sh
```
Comment (brandond) [inline Dockerfile:109]: These need to be bumped to the actual version merged to rke2-charts. They should also not have the -rc00 suffix on them.

File: scripts/build-images [hunk 4]
```diff
@@ -71,5 +71,14 @@ xargs -n1 -t docker image pull --quiet << EOF > build/images-multus.txt
     ${REGISTRY}/rancher/hardened-sriov-network-webhook:v1.0.0-build20210429
 EOF
 
+xargs -n1 -t docker image pull --quiet << EOF > build/images-harvester.txt
+    ${REGISTRY}/rancher/harvester-cloud-provider:v0.1.1
+    ${REGISTRY}/rancher/harvester-csi-driver:v0.1.1
+    ${REGISTRY}/rancher/longhornio-csi-node-driver-registrar:v2.3.0
+    ${REGISTRY}/rancher/longhornio-csi-resizer:v1.2.0
+    ${REGISTRY}/rancher/longhornio-csi-provisioner:v2.1.2
+    ${REGISTRY}/rancher/longhornio-csi-attacher:v3.2.1
```
Comment (brandond) [inline scripts/build-images:80]: did any of these get bumped in the time this has been waiting for review?

**Review by brandond — CHANGES_REQUESTED** [review]

File: scripts/build-images (hunk 4 above)
Comment (yaocw2020) [reply scripts/build-images:80]: Yes, we have corrected them to the latest version.

**Review by brandond — APPROVED** [review]

**Review by briandowns — APPROVED** [review]

**Pull request merged**